utils: fix snap environment detection and add tests (CRAFT-376)

cmatsuoka opened the pull request:
Snap environment detection returned false positives and was not being
properly tested. Fix snap environment detection and add tests for
`is_snap()` and `is_inside_container()`.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

cjp256 commented on craft_parts/utils/os_utils.py:
```diff
@@ -172,13 +172,9 @@ def is_snap(*, application_name: Optional[str] = None) -> bool:
         res = snap_name == application_name
     else:
-        res = snap_name is not None
+        res = snap_name != ""
```
Generally i'd suggest using the default of None to distinguish if envvar is actually set to empty string.

```
$ FOO="" python -c 'import os; print(repr(os.environ.get("FOO")))'
''
$ python -c 'import os; print(os.environ.get("FOO"))'
None
```

But in this case it's non-consequential, so I'm ok as-is.

cmatsuoka replied:
I used the empty string to avoid an extra test before comparing to `application_name`.

sergiusens replied:
what if SNAP_NAME is set but does not match application name?

- Launch emacs/vscode snap
- start vterm
- env will have SNAP_NAME, but not the SNAP_NAME we want

we don't care if SNAP_NAME is set, only from the point of view of it matching the application name, I believe my simpler check does the trick.

cmatsuoka replied:
This is an interesting issue because craft-parts doesn't know the snap name of the application it's running under, so perhaps it should check if SNAP_NAME is set to something, regardless of the name.

sergiusens commented on craft_parts/utils/os_utils.py:
```diff
@@ -172,13 +172,9 @@ def is_snap(*, application_name: Optional[str] = None) -> bool:
     if application_name:
         res = snap_name == application_name
```
Does this not work?
```suggestion
    is_snap = application_name == snap_name:
```

sergiusens commented on craft_parts/utils/os_utils.py:
```diff
@@ -172,13 +172,9 @@ def is_snap(*, application_name: Optional[str] = None) -> bool:
-        res,
-        snap_name,
-    )
+    logger.debug("is_snap: %s, SNAP_NAME set to %s", res, snap_name)
```
```suggestion
    logger.debug("is_snap: %s, SNAP_NAME set to %s ", is_snap, snap_name)
```

sergiusens commented:
Closing in favor of #99

cmatsuoka commented:
Updated to use `application_name` to check if we're running as a snap. A few notes on this change:
- Updated `is_snap()` to require an application name passed, it's not optional anymore.
- Changed package manager configuration to be executed only once (when the lifecycle manager is created).
- Added an optional `application_package_name` parameter which can be used if `application_name` can't match the snap name.
